### PR TITLE
Range

### LINKED ISSUE #9: Create Trigger with "total_count"
A trigger is needed that will run until a "total_count" is reached and then will stop completely.

### PR DESCRIPTION
**Pull request by cloudformdesign**
fixes #9 

Create `RangeFastSimulator` to do qualification testing

Some notes:
- the `RangeGenerator` returns a generator object, not a list. This is intentional, and I think every simulator should do this if able.
- `LimitTrigger` has similar functionality to `SimulateFast` -- only one thread can run at a time and it will only generate `total_signals` number of signals.

### CODE REVIEW

File: blocks.py [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
+from .generators.range import RangeGenerator
 from .generators.identity import IdentityGenerator
 from .triggers.interval import IntervalTrigger
+from .triggers.limit import LimitTrigger
 from .triggers.safe import SafeTrigger
 from .triggers.fast import FastTrigger
 from nio.common.block.base import Block
 from nio.common.discovery import Discoverable, DiscoverableType
-from nio.metadata.properties import VersionProperty
+from nio.metadata.properties.version import VersionProperty
```
Comment (cloudformdesign) [inline blocks.py]: this did not import without this change (nio 1.5.1)

**Comment by cloudformdesign** [comment]
Also, `RangeGenerator` is not currently thread safe. A lock would be alright, I suppose, but not ideal. Thoughts?

**Comment by mattdodge** [comment]
These look like good improvements, here are a couple suggestions though, mainly surrounding the way we split this up and approach it.

1. It seems that the new `RangeGenerator` serves the same purpose as the `CounterGenerator`, only it performs better. If this is the case, we should make a separate pull request to modify the existing CounterGenerator to use this method.
2. Similarly, the `LimitTrigger` seems like the same thing as the `FastTrigger`, only with a configurable batch size. Just like the first item, we should make a separate PR updating the `FastTrigger` to work this way.
3. If either of these new components remain new, we should include their function in the README. See the existing sections for examples of how to document them.

File: blocks.py (hunk 1 above)
Comment (mattdodge) [inline blocks.py]: Good call, thanks. This will be enabled in `1.5.2` but we should use `.version` for now

**Comment by mattdodge** [comment]
Regarding the generators vs list topic, I'm ok with changing the return type of `generate_signals` to return a generator. However, we should not be notifying generators using `notify_signals`. I know that it works in some (most?) cases, but it is not officially supported. So I think a change to the return type of the function would change the way the triggers would call it also. I would be interested in seeing the implementation keeping these things in mind.

**Comment by cloudformdesign** [comment]
- `RangeGenerator` will perform different from `CounterGenerator` if there is excessive load and multiple threads trying to be run. `CounterGenerator` will return something like:
```
[1  3  5  7  9 ...
    [3  6  8  10 ... ]
```
Where as `RangeGenerator` will do:
```
[1  2  3  4  5  ...]
                         [1   2   3   4   5  ...]
```

- The major difference with `Limit` and `Fast` is:
  - batch size is configurable
  - `Fast` doesn't have the ability to configure a period. This confused me a lot, was it intended?

- See `LimitTrigger` for an implementation of generators working side/by/side with lists as the output: https://github.com/nio-blocks/simulator/blob/range/triggers/limit.py#L48

**Comment by mattdodge** [comment]
This is covered by other PRs now, closing

**Pull request closed without merging**